File: examples/go/inspect.go
package main

import (
	"fmt"

	tree_sitter "github.com/tree-sitter/go-tree-sitter"
	tree_sitter_gram "github.com/tree-sitter/tree-sitter-gram/bindings/go"
)

func main() {
	parser := tree_sitter.NewParser()
	parser.SetLanguage(tree_sitter.NewLanguage(tree_sitter_gram.Language()))

	sourceCode := []byte(`(hello:Greeting {text: "hi"})-[:TO]->(world:Place)`)
	tree := parser.Parse(sourceCode, nil)

	fmt.Println("Parse tree:")
	fmt.Println(tree.RootNode().ToSexp())

	nodes := 0
	relationships := 0
	var walk func(*tree_sitter.Node)
	walk = func(node *tree_sitter.Node) {
		switch node.Kind() {
		case "node_pattern":
			nodes++
			fmt.Printf("  node: %s\n", node.Utf8Text(sourceCode))
		case "relationship_pattern":
			relationships++
			fmt.Printf("  relationship: %s\n", node.Utf8Text(sourceCode))
		}
		for i := uint(0); i < node.ChildCount(); i++ {
			walk(node.Child(i))
		}
	}
	walk(tree.RootNode())

	fmt.Printf("\nSummary: %d nodes, %d relationships\n", nodes, relationships)
}

File: examples/node/inspect.js
const Parser = require("tree-sitter");
const GramLang = require("tree-sitter-gram");

const parser = new Parser();
parser.setLanguage(GramLang);

const gramCode = `(hello:Greeting {text: "hi"})-[:TO]->(world:Place {name: "Earth"})
(alice:Person)-[:KNOWS]->(bob:Person)`;

const tree = parser.parse(gramCode);

console.log("Parse tree:");
console.log(tree.rootNode.toString());
console.log();

function walk(node, source, stats) {
  switch (node.type) {
    case "node":
      stats.nodes.push(source.slice(node.startIndex, node.endIndex));
      break;
    case "relationship":
      stats.relationships.push(source.slice(node.startIndex, node.endIndex));
      break;
  }
  for (let i = 0; i < node.childCount; i++) {
    walk(node.child(i), source, stats);
  }
}

const stats = { nodes: [], relationships: [] };
walk(tree.rootNode, gramCode, stats);

for (const n of stats.nodes) {
  console.log("  node:", n);
}
for (const r of stats.relationships) {
  console.log("  relationship:", r);
}

console.log(`\nSummary: ${stats.nodes.length} nodes, ${stats.relationships.length} relationships`);
